### PR TITLE
evmc tool CLI improvements

### PR DESCRIPTION
**Pull request by chfast**
For arguments `code` and `--input` we try to guess if the value is a file path or hex string. If it is a file then we load hex string from the file.

There are some gaps here still: we assume the file contains valid hex string. We could additionally inspect the file content but that requires custom `ExistingFile` validator.

Closes https://github.com/ethereum/evmc/issues/569.

### CODE REVIEW

File: tools/evmc/main.cpp [hunk 1]
```diff
@@ -3,29 +3,84 @@
...
+        return evmc::from_hex(str);
+
+    // Must be a file path.
+    std::ifstream file{str, std::ios::in | std::ios::binary};
+    const auto content =
+        std::string(std::istreambuf_iterator<char>{file}, std::istreambuf_iterator<char>{});
+    return evmc::from_hex(content);
+}
+
+struct HexValidator : public CLI::Validator
```
Comment (axic) [inline tools/evmc/main.cpp:38]: Nice!

File: tools/evmc/main.cpp [hunk 2]
```diff
@@ -3,29 +3,84 @@
...
+    auto create = false;
 
     CLI::App app{"EVMC tool"};
     const auto& version_flag = *app.add_flag("--version", "Print version information and exit");
     const auto& vm_option =
         *app.add_option("--vm", vm_config, "EVMC VM module")->envname("EVMC_VM");
 
     auto& run_cmd = *app.add_subcommand("run", "Execute EVM bytecode");
-    run_cmd.add_option("code", code_hex, "Hex-encoded bytecode")->required();
+    run_cmd.add_option("code", code_hex, "Bytecode")->required()->check(Hex | CLI::ExistingFile);
```
Comment (axic) [inline tools/evmc/main.cpp:77]: Cool!

File: tools/evmc/main.cpp [hunk 3]
```diff
@@ -23,10 +43,10 @@ int main(int argc, const char** argv)
         *app.add_option("--vm", vm_config, "EVMC VM module")->envname("EVMC_VM");
 
     auto& run_cmd = *app.add_subcommand("run", "Execute EVM bytecode");
-    run_cmd.add_option("code", code_hex, "Hex-encoded bytecode")->required();
+    run_cmd.add_option("code", code_hex, "Bytecode")->required()->check(Hex);
```
Comment (axic) [inline tools/evmc/main.cpp]: So this replaces `required`, does that mean HexValidator should fail if the input is shorter than 2 nibbles?

Comment (axic) [reply tools/evmc/main.cpp]: This also raises the question, is there a CLI test for calling this without input bytecode?

File: tools/evmc/main.cpp [hunk 4]
```diff
@@ -6,9 +6,24 @@
 #include <CLI/CLI.hpp>
 #include <evmc/hex.hpp>
 #include <evmc/loader.h>
+#include <fstream>
 
 namespace
 {
+/// Returns the input str if already valid hex string. Otherwise, interprets the str as a file
+/// name and loads the hex string from this file.
+/// @todo The file content is not validated.
```
Comment (axic) [inline tools/evmc/main.cpp:15]: So does the file needs to be a hex string, or it is a binary file?

File: tools/evmc/main.cpp (hunk 3 above)
Comment (chfast) [reply tools/evmc/main.cpp]: The "required" mean the argument is required. Values are always validated. Any order works.

File: tools/evmc/main.cpp (hunk 4 above)
Comment (chfast) [reply tools/evmc/main.cpp:15]: hex string

File: tools/evmc/main.cpp (hunk 3 above)
Comment (axic) [reply tools/evmc/main.cpp]: Oh, for some reason the split diff didn't show it to be already chained. So just ignore this question then.

File: tools/evmc/main.cpp (hunk 4 above)
Comment (axic) [reply tools/evmc/main.cpp:15]: Can it be validated? Even if just validating it here with `abort`, I know it is not nice, but reduces confusion should someone misuse it.

Comment (chfast) [reply tools/evmc/main.cpp:15]: It still throws "invalid hex digit" exception or similar. This is caught and reported although it is not clear where the error comes from. I added test for this case.

Comment (axic) [reply tools/evmc/main.cpp:15]: Isn't it coming from somewhere inside `cmd::run`?

Comment (chfast) [reply tools/evmc/main.cpp:15]: It is. But I mean user will see the error about invalid hex without information about the source. He has to guess it is code and/or input what is invalid.

Comment (axic) [reply tools/evmc/main.cpp:15]: I think it comes from there as it uses `load_hex` which throws the error and `main` catches that. Can you leave this as a comment here (that it is caught later in `evmc::run`), for now?

File: test/tools/CMakeLists.txt [hunk 5]
```diff
@@ -39,5 +39,35 @@ add_evmc_tool_test(
...
 
+add_evmc_tool_test(
+    invalid_hex_code
+    "--vm $<TARGET_FILE:evmc::example-vm> run 0x600"
+    "code: \\(incomplete hex byte pair\\) OR \\(File does not exist: 0x600\\)"
+)
+
+add_evmc_tool_test(
+    invalid_hex_input
+    "--vm $<TARGET_FILE:evmc::example-vm> run 0x --input aa0y"
```
Comment (axic) [inline test/tools/CMakeLists.txt:50]: Hmm, we accept this weird Ethereum concept of `0x` for empty input? Does it accept `run ''` or `run ""` ?

Comment (chfast) [reply test/tools/CMakeLists.txt:50]: All should work

Comment (axic) [reply test/tools/CMakeLists.txt:50]: Could you add one more test with that? I hope we can remove the `run 0x` case at some point in the future :)

Comment (chfast) [reply test/tools/CMakeLists.txt:50]: https://github.com/ethereum/evmc/blob/master/test/unittests/hex_test.cpp#L71

File: test/tools/CMakeLists.txt [hunk 6]
```diff
@@ -39,5 +39,35 @@ add_evmc_tool_test(
     "Result: +success[\r\n]+Gas used: +6[\r\n]+Output: +02[\r\n]"
 )
 
+add_evmc_tool_test(
+    invalid_hex_code
+    "--vm $<TARGET_FILE:evmc::example-vm> run 0x600"
+    "code: \\(incomplete hex byte pair\\) OR \\(File does not exist: 0x600\\)"
```
[the remark this answers is not in the document]
Comment (chfast) [reply test/tools/CMakeLists.txt:45]: This is CLI11 default but probably can be changed if you re-implement many things. I'm not going to do it now.

**Review by gumb0 — APPROVED** [review]

File: test/tools/CMakeLists.txt (hunk 5 above)
Comment (axic) [reply test/tools/CMakeLists.txt:50]: I meant adding a test for `run ""`.

And yes, I know it currently accepts `0x` as empty and not asking to remove it right now :)

Comment (chfast) [reply test/tools/CMakeLists.txt:50]: I tried 30 mins, but I don't know how to do this in CMake.

Comment (chfast) [reply test/tools/CMakeLists.txt:56]: Added.

Comment (axic) [reply test/tools/CMakeLists.txt:56]: Hmm, I thought `"--vm $<TARGET_FILE:evmc::example-vm> run '' --input aa0y"` would have been enough?

**Review by axic — APPROVED** [review]

**Pull request merged**